Configurable tool versions should probably have a hardcoded constraint range.

Any tool Pants uses must have the CLI syntax Pants expects when Pants is released. If a user works around a tool issue by configuring the tool version, that can leave a custom tool configuration that works at the moment, but fails mysteriously when Pants is upgraded to a version that expects the tool to be able to support new arguments. We should probably support a floor for this sort of case and at least log a warning. More generally we probably need to support a range.

This surfaced in the wild with a user that had configured the following when migrating from Pants 2.2 to 2.3:
```toml
[download-pex-bin]
version = "v2.1.33"
known_versions = [
    "v2.1.33|darwin|7f9f6168691fe83f38fbc248bd8629a152cf2a8833be1afdc06219f70fbb6064|3596348",
    "v2.1.33|linux|7f9f6168691fe83f38fbc248bd8629a152cf2a8833be1afdc06219f70fbb6064|3596348"
]
```
Upon upgrade to Pants 2.5.x which ships using Pex 2.1.42 they hit mysterious Pex errors:
```
21:27:43.95 [INFO] Starting: Resolving plugins: toolchain.pants.plugin==0.11.0
21:27:44.31 [INFO] Completed: Resolving plugins: toolchain.pants.plugin==0.11.0
21:27:44.31 [ERROR] 1 Exception encountered:
  ProcessExecutionFailure: Process 'Resolving plugins: toolchain.pants.plugin==0.11.0' failed with exit code 2.
stdout:
stderr:
usage: pex [-o OUTPUT.PEX] [options] [-- arg1 arg2 ...]
pex builds a PEX (Python Executable) file based on the given specifications: sources, requirements, their dependencies and other options.
pex: error: unrecognized arguments: toolchain.pants.plugin==0.11.0
```

This error, in the end, is because Pex 2.1.33 does not understand arguments added in Pex 2.1.42.

--- a/src/python/pants/backend/python/util_rules/pex_cli.py
+++ b/src/python/pants/backend/python/util_rules/pex_cli.py
@@ -39,6 +39,7 @@ class PexBinary(TemplatedExternalTool):
 
     default_version = "v2.1.42"
     default_url_template = "https://github.com/pantsbuild/pex/releases/download/{version}/pex"
+    version_constraints = ">=2.1.42, <3.0"
 
     @classproperty
     def default_known_versions(cls):

--- a/src/python/pants/core/util_rules/external_tool.py
+++ b/src/python/pants/core/util_rules/external_tool.py
@@ -35,16 +35,15 @@ class UnsupportedVersion(ExternalToolError):
     constraints."""
 
 
-class ExternalToolVersionConstraintsViolationAction(Enum):
-    """What action to take in case the requested version does not satisfy the version constraints
-    for this tool.
+class UnsupportedVersionUsage(Enum):
+    """What action to take in case the requested version of {tool} is not supported.
 
-    Current constraints: {constraints}
+    Supported versions: {constraints}
     """
 
-    LogWarning = "warning"
+    Allow = "allow"
     RaiseError = "error"
-    Ignore = "ignore"
+    LogWarning = "warning"
 
 
 @dataclass(frozen=True)
@@ -124,7 +123,9 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
             default=cls.default_version,
             advanced=True,
             help=f"Use this version of {cls.name}."
-            + (f" Constraints {cls.version_constraints}" if cls.version_constraints else ""),
+            + (
+                f" Supported versions: {cls.version_constraints}" if cls.version_constraints else ""
+            ),
         )
 
         help_str = textwrap.dedent(
@@ -157,16 +158,17 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
         )
 
         register(
-            "--on-version-constraints-violation",
+            "--use-unsupported-version",
             advanced=True,
-            type=ExternalToolVersionConstraintsViolationAction,
+            type=UnsupportedVersionUsage,
             help=textwrap.dedent(
                 "    "
-                + ExternalToolVersionConstraintsViolationAction.__doc__.format(
-                    constraints=cls.version_constraints
+                + UnsupportedVersionUsage.__doc__.format(
+                    tool=cls.name,
+                    constraints=cls.version_constraints if cls.version_constraints else "any",
                 )
             ),
-            default=ExternalToolVersionConstraintsViolationAction.LogWarning,
+            default=UnsupportedVersionUsage.LogWarning,
         )
 
     @property
@@ -229,22 +231,22 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
             # all ok
             return None
 
-        action = self.options.on_version_constraints_violation
-        if action is ExternalToolVersionConstraintsViolationAction.Ignore:
+        action = self.options.use_unsupported_version
+        if action is UnsupportedVersionUsage.Allow:
             logger.debug(
                 f"Ignoring constraints for version {version}, does not satisfy version constraints {self.constraints}"
             )
-        elif action is ExternalToolVersionConstraintsViolationAction.LogWarning:
+        elif action is UnsupportedVersionUsage.LogWarning:
             logger.warning(
                 f"{self.name} {version} may not be compatible with this version of pants, as it does not satisfy the constraints {self.constraints}"
             )
-        elif action is ExternalToolVersionConstraintsViolationAction.RaiseError:
+        elif action is UnsupportedVersionUsage.RaiseError:
             raise UnsupportedVersion(
-                f"Version {version} does not satisfy the version constraints {self.constraints}"
+                f"{self.name} {version} does not satisfy the version constraints {self.constraints}"
             )
         else:
             raise ValueError(
-                f"Internal Error: unexpected value for --on-version-constraints-violation: {action!r}"
+                f"Internal Error: unexpected value for --use-unsupported-version: {action!r}"
             )
 
 

--- a/src/python/pants/core/util_rules/external_tool.py
+++ b/src/python/pants/core/util_rules/external_tool.py
@@ -6,6 +6,8 @@ from abc import ABCMeta, abstractmethod
 from dataclasses import dataclass
 from typing import Dict, List, Optional, Tuple, cast
 
+from pkg_resources import Requirement
+
 from pants.core.util_rules import archive
 from pants.core.util_rules.archive import ExtractedArchive
 from pants.engine.fs import Digest, DownloadFile, FileDigest
@@ -22,6 +24,11 @@ class UnknownVersion(Exception):
 
 class ExternalToolError(Exception):
     pass
+
+
+class UnsupportedVersion(ExternalToolError):
+    """The specified version of the tool is not supported, according to the given version
+    constraints."""
 
 
 @dataclass(frozen=True)
@@ -52,6 +59,8 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
           "1.2.3|linux |1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd|333333",
         ]
 
+        version_constraints = ">=1.2.3, <2.0"
+
         def generate_url(self, plat: Platform) -> str:
             ...
 
@@ -73,6 +82,15 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
     default_version: str
     default_known_versions: List[str]
 
+    # our hard constraints required for this tool, to support the API we expect from it
+    version_constraints: Optional[str] = None
+    constraints: Optional[Requirement] = None
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.version_constraints:
+            self.constraints = Requirement.parse(f"{self.name}{self.version_constraints}")
+
     @classproperty
     def name(cls):
         """The name of the tool, for use in user-facing messages.
@@ -89,7 +107,12 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
             type=str,
             default=cls.default_version,
             advanced=True,
-            help=f"Use this version of {cls.name}.",
+            help=f"Use this version of {cls.name}."
+            + (
+                f" Must satisfy constraints {cls.version_constraints}"
+                if cls.version_constraints
+                else ""
+            ),
         )
 
         help_str = textwrap.dedent(
@@ -152,6 +175,7 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
 
     def get_request(self, plat: Platform) -> ExternalToolRequest:
         """Generate a request for this tool."""
+        self.check_version_constraint(self.version)
         for known_version in self.known_versions:
             try:
                 ver, plat_val, sha256, length = (x.strip() for x in known_version.split("|"))
@@ -174,6 +198,12 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
             f"No known version of {self.name} {self.version} for {plat.value} found in "
             f"{self.known_versions}"
         )
+
+    def check_version_constraint(self, version: str) -> None:
+        if self.constraints and not self.constraints.specifier.contains(version):  # type: ignore[attr-defined]
+            raise UnsupportedVersion(
+                f"Version {version} does not satisfy the version constraint {self.constraints}"
+            )
 
 
 class TemplatedExternalTool(ExternalTool):

--- a/src/python/pants/core/util_rules/external_tool.py
+++ b/src/python/pants/core/util_rules/external_tool.py
@@ -1,9 +1,11 @@
 # Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+import logging
 import textwrap
 from abc import ABCMeta, abstractmethod
 from dataclasses import dataclass
+from enum import Enum
 from typing import Dict, List, Optional, Tuple, cast
 
 from pkg_resources import Requirement
@@ -17,6 +19,8 @@ from pants.option.subsystem import Subsystem
 from pants.util.logging import LogLevel
 from pants.util.meta import classproperty
 
+logger = logging.getLogger(__name__)
+
 
 class UnknownVersion(Exception):
     pass
@@ -29,6 +33,18 @@ class ExternalToolError(Exception):
 class UnsupportedVersion(ExternalToolError):
     """The specified version of the tool is not supported, according to the given version
     constraints."""
+
+
+class ExternalToolVersionConstraintsViolationAction(Enum):
+    """What action to take in case the requested version does not satisfy the version constraints
+    for this tool.
+
+    Current constraints: {constraints}
+    """
+
+    LogWarning = "warning"
+    RaiseError = "error"
+    Ignore = "ignore"
 
 
 @dataclass(frozen=True)
@@ -108,11 +124,7 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
             default=cls.default_version,
             advanced=True,
             help=f"Use this version of {cls.name}."
-            + (
-                f" Must satisfy constraints {cls.version_constraints}"
-                if cls.version_constraints
-                else ""
-            ),
+            + (f" Constraints {cls.version_constraints}" if cls.version_constraints else ""),
         )
 
         help_str = textwrap.dedent(
@@ -142,6 +154,19 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
             default=cls.default_known_versions,
             advanced=True,
             help=help_str,
+        )
+
+        register(
+            "--on-version-constraints-violation",
+            advanced=True,
+            type=ExternalToolVersionConstraintsViolationAction,
+            help=textwrap.dedent(
+                "    "
+                + ExternalToolVersionConstraintsViolationAction.__doc__.format(
+                    constraints=cls.version_constraints
+                )
+            ),
+            default=ExternalToolVersionConstraintsViolationAction.LogWarning,
         )
 
     @property
@@ -200,9 +225,26 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
         )
 
     def check_version_constraint(self, version: str) -> None:
-        if self.constraints and not self.constraints.specifier.contains(version):  # type: ignore[attr-defined]
+        if not self.constraints or self.constraints.specifier.contains(version):  # type: ignore[attr-defined]
+            # all ok
+            return None
+
+        action = self.options.on_version_constraints_violation
+        if action is ExternalToolVersionConstraintsViolationAction.Ignore:
+            logger.debug(
+                f"Ignoring constraints for version {version}, does not satisfy version constraints {self.constraints}"
+            )
+        elif action is ExternalToolVersionConstraintsViolationAction.LogWarning:
+            logger.warning(
+                f"{self.name} {version} may not be compatible with this version of pants, as it does not satisfy the constraints {self.constraints}"
+            )
+        elif action is ExternalToolVersionConstraintsViolationAction.RaiseError:
             raise UnsupportedVersion(
-                f"Version {version} does not satisfy the version constraint {self.constraints}"
+                f"Version {version} does not satisfy the version constraints {self.constraints}"
+            )
+        else:
+            raise ValueError(
+                f"Internal Error: unexpected value for --on-version-constraints-violation: {action!r}"
             )
 
 

--- a/src/python/pants/core/util_rules/external_tool.py
+++ b/src/python/pants/core/util_rules/external_tool.py
@@ -36,10 +36,7 @@ class UnsupportedVersion(ExternalToolError):
 
 
 class UnsupportedVersionUsage(Enum):
-    """What action to take in case the requested version of {tool} is not supported.
-
-    Supported versions: {constraints}
-    """
+    """What action to take in case the requested version of the tool is not supported."""
 
     Allow = "allow"
     RaiseError = "error"
@@ -92,19 +89,19 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
         ...
     """
 
-    # The default values for --version and --known-versions.
+    # The default values for --version and --known-versions, and the supported versions.
     # Subclasses must set appropriately.
     default_version: str
     default_known_versions: List[str]
-
-    # our hard constraints required for this tool, to support the API we expect from it
     version_constraints: Optional[str] = None
-    constraints: Optional[Requirement] = None
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         if self.version_constraints:
-            self.constraints = Requirement.parse(f"{self.name}{self.version_constraints}")
+            constraints = Requirement.parse(f"{self.name}{self.version_constraints}")
+            self.check_version_constraints(
+                self.version, constraints, self.options.use_unsupported_version
+            )
 
     @classproperty
     def name(cls):
@@ -124,7 +121,9 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
             advanced=True,
             help=f"Use this version of {cls.name}."
             + (
-                f" Supported versions: {cls.version_constraints}" if cls.version_constraints else ""
+                f"\n\nSupported {cls.name} versions: {cls.version_constraints}"
+                if cls.version_constraints
+                else ""
             ),
         )
 
@@ -162,13 +161,13 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
             advanced=True,
             type=UnsupportedVersionUsage,
             help=textwrap.dedent(
-                "    "
-                + UnsupportedVersionUsage.__doc__.format(
-                    tool=cls.name,
-                    constraints=cls.version_constraints if cls.version_constraints else "any",
-                )
+                f"""
+                What action to take in case the requested version of {cls.name} is not supported.
+
+                Supported {cls.name} versions: {cls.version_constraints if cls.version_constraints else "unspecified"}
+                """
             ),
-            default=UnsupportedVersionUsage.LogWarning,
+            default=UnsupportedVersionUsage.RaiseError,
         )
 
     @property
@@ -202,7 +201,6 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
 
     def get_request(self, plat: Platform) -> ExternalToolRequest:
         """Generate a request for this tool."""
-        self.check_version_constraint(self.version)
         for known_version in self.known_versions:
             try:
                 ver, plat_val, sha256, length = (x.strip() for x in known_version.split("|"))
@@ -226,28 +224,34 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
             f"{self.known_versions}"
         )
 
-    def check_version_constraint(self, version: str) -> None:
-        if not self.constraints or self.constraints.specifier.contains(version):  # type: ignore[attr-defined]
+    @classmethod
+    def check_version_constraints(
+        cls, version: str, constraints: Requirement, action: UnsupportedVersionUsage
+    ) -> None:
+        if constraints.specifier.contains(version):  # type: ignore[attr-defined]
             # all ok
             return None
 
-        action = self.options.use_unsupported_version
+        msg = [
+            f"The option [{cls.options_scope}].version is set to {version}, which is not compatible with what this release of Pants expects: {constraints}",
+        ]
+
         if action is UnsupportedVersionUsage.Allow:
-            logger.debug(
-                f"Ignoring constraints for version {version}, does not satisfy version constraints {self.constraints}"
+            msg.append(
+                f"The option [{cls.options_scope}].use_unsupported_version is set to 'allow'."
             )
-        elif action is UnsupportedVersionUsage.LogWarning:
-            logger.warning(
-                f"{self.name} {version} may not be compatible with this version of pants, as it does not satisfy the constraints {self.constraints}"
-            )
-        elif action is UnsupportedVersionUsage.RaiseError:
-            raise UnsupportedVersion(
-                f"{self.name} {version} does not satisfy the version constraints {self.constraints}"
-            )
+            logger.info("\n".join(msg))
         else:
-            raise ValueError(
-                f"Internal Error: unexpected value for --use-unsupported-version: {action!r}"
+            msg.append(
+                "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version."
             )
+            if action is UnsupportedVersionUsage.LogWarning:
+                logger.warning("\n".join(msg))
+            elif action is UnsupportedVersionUsage.RaiseError:
+                msg.append(
+                    f"Alternatively, update [{cls.options_scope}].use_unsupported_version to be 'allow' or 'warning'."
+                )
+                raise UnsupportedVersion("\n".join(msg))
 
 
 class TemplatedExternalTool(ExternalTool):

--- a/src/python/pants/core/util_rules/external_tool.py
+++ b/src/python/pants/core/util_rules/external_tool.py
@@ -38,7 +38,6 @@ class UnsupportedVersion(ExternalToolError):
 class UnsupportedVersionUsage(Enum):
     """What action to take in case the requested version of the tool is not supported."""
 
-    Allow = "allow"
     RaiseError = "error"
     LogWarning = "warning"
 
@@ -233,25 +232,26 @@ class ExternalTool(Subsystem, metaclass=ABCMeta):
             return None
 
         msg = [
-            f"The option [{cls.options_scope}].version is set to {version}, which is not compatible with what this release of Pants expects: {constraints}",
+            f"The option [{cls.options_scope}].version is set to {version}, which is not compatible",
+            f"with what this release of Pants expects: {constraints}.",
+            "Please update the version to a supported value, or consider using a different Pants",
+            "release if you cannot change the version.",
         ]
 
-        if action is UnsupportedVersionUsage.Allow:
-            msg.append(
-                f"The option [{cls.options_scope}].use_unsupported_version is set to 'allow'."
+        if action is UnsupportedVersionUsage.LogWarning:
+            msg.extend(
+                [
+                    "Alternatively, you can ignore this warning (at your own peril) by adding this",
+                    "to the GLOBAL section of pants.toml:",
+                    f"""ignore_warnings = ["The option [{cls.options_scope}].version is set to"].""",
+                ]
             )
-            logger.info("\n".join(msg))
-        else:
+            logger.warning(" ".join(msg))
+        elif action is UnsupportedVersionUsage.RaiseError:
             msg.append(
-                "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version."
+                f"Alternatively, update [{cls.options_scope}].use_unsupported_version to be 'warning'."
             )
-            if action is UnsupportedVersionUsage.LogWarning:
-                logger.warning("\n".join(msg))
-            elif action is UnsupportedVersionUsage.RaiseError:
-                msg.append(
-                    f"Alternatively, update [{cls.options_scope}].use_unsupported_version to be 'allow' or 'warning'."
-                )
-                raise UnsupportedVersion("\n".join(msg))
+            raise UnsupportedVersion(" ".join(msg))
 
 
 class TemplatedExternalTool(ExternalTool):

--- a/src/python/pants/core/util_rules/external_tool_test.py
+++ b/src/python/pants/core/util_rules/external_tool_test.py
@@ -10,10 +10,10 @@ from pants.core.util_rules.external_tool import (
     ExternalTool,
     ExternalToolError,
     ExternalToolRequest,
-    ExternalToolVersionConstraintsViolationAction,
     TemplatedExternalTool,
     UnknownVersion,
     UnsupportedVersion,
+    UnsupportedVersionUsage,
 )
 from pants.engine.fs import DownloadFile, FileDigest
 from pants.engine.platform import Platform
@@ -144,16 +144,16 @@ def no_exception():
     [
         (
             "v1.2.3",
-            ExternalToolVersionConstraintsViolationAction.RaiseError,
+            UnsupportedVersionUsage.RaiseError,
             pytest.raises(
                 UnsupportedVersion,
-                match="Version v1.2.3 does not satisfy the version constraints foobar<3.8,>3.2.1",
+                match="foobar v1.2.3 does not satisfy the version constraints foobar<3.8,>3.2.1",
             ),
             None,
         ),
         (
             "v3.2.2",
-            ExternalToolVersionConstraintsViolationAction.RaiseError,
+            UnsupportedVersionUsage.RaiseError,
             pytest.raises(
                 UnknownVersion, match="No known version of foobar v3.2.2 for darwin found in"
             ),
@@ -161,21 +161,21 @@ def no_exception():
         ),
         (
             "v3.4.7",
-            ExternalToolVersionConstraintsViolationAction.RaiseError,
+            UnsupportedVersionUsage.RaiseError,
             no_exception(),
             None,
         ),
         (
             "v3.8.0",
-            ExternalToolVersionConstraintsViolationAction.RaiseError,
+            UnsupportedVersionUsage.RaiseError,
             pytest.raises(
-                UnsupportedVersion, match="Version v3.8.0 does not satisfy the version constraints"
+                UnsupportedVersion, match="foobar v3.8.0 does not satisfy the version constraints"
             ),
             None,
         ),
         (
             "v3.8.0",
-            ExternalToolVersionConstraintsViolationAction.LogWarning,
+            UnsupportedVersionUsage.LogWarning,
             pytest.raises(
                 UnknownVersion, match="No known version of foobar v3.8.0 for darwin found in"
             ),
@@ -188,7 +188,7 @@ def no_exception():
         ),
         (
             "v3.8.0",
-            ExternalToolVersionConstraintsViolationAction.Ignore,
+            UnsupportedVersionUsage.Allow,
             pytest.raises(
                 UnknownVersion, match="No known version of foobar v3.8.0 for darwin found in"
             ),
@@ -209,7 +209,7 @@ def test_version_constraints(caplog, version, action, assert_expectation, expect
         create_subsystem(
             ConstrainedTool,
             version=version,
-            on_version_constraints_violation=action,
+            use_unsupported_version=action,
             known_versions=ConstrainedTool.default_known_versions,
             url_template=ConstrainedTool.default_url_template,
             url_platform_mapping=ConstrainedTool.default_url_platform_mapping,

--- a/src/python/pants/core/util_rules/external_tool_test.py
+++ b/src/python/pants/core/util_rules/external_tool_test.py
@@ -1,6 +1,7 @@
 # Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+import logging
 from contextlib import contextmanager
 
 import pytest
@@ -9,6 +10,7 @@ from pants.core.util_rules.external_tool import (
     ExternalTool,
     ExternalToolError,
     ExternalToolRequest,
+    ExternalToolVersionConstraintsViolationAction,
     TemplatedExternalTool,
     UnknownVersion,
     UnsupportedVersion,
@@ -114,13 +116,13 @@ class ConstrainedTool(TemplatedExternalTool):
     name = "foobar"
     options_scope = "foobar"
     version_constraints = ">3.2.1, <3.8"
-    default_version = "3.4.7"
+    default_version = "v3.4.7"
     default_known_versions = [
-        "3.2.0|darwin   |1102324cdaacd589e50b8b7770595f220f54e18a1d76ee3c445198f80ab865b8|123346",
-        "3.2.0|linux_ppc|39e5d64b0f31117c94651c880d0a776159e49eab42b2066219569934b936a5e7|124443",
-        "3.2.0|linux    |c0c667fb679a8221bed01bffeed1f80727c6c7827d0cbd8f162195efb12df9e4|121212",
-        "3.4.7|darwin   |9d0e18cd74b918c7b3edd0203e75569e0c8caecb1367b3be409b45e28514f5be|123321",
-        "3.4.7|linux    |a019dfc4b32d63c1392aa264aed2253c1e0c2fb09216f8e2cc269bbfb8bb49b5|134213",
+        "v3.2.0|darwin   |1102324cdaacd589e50b8b7770595f220f54e18a1d76ee3c445198f80ab865b8|123346",
+        "v3.2.0|linux_ppc|39e5d64b0f31117c94651c880d0a776159e49eab42b2066219569934b936a5e7|124443",
+        "v3.2.0|linux    |c0c667fb679a8221bed01bffeed1f80727c6c7827d0cbd8f162195efb12df9e4|121212",
+        "v3.4.7|darwin   |9d0e18cd74b918c7b3edd0203e75569e0c8caecb1367b3be409b45e28514f5be|123321",
+        "v3.4.7|linux    |a019dfc4b32d63c1392aa264aed2253c1e0c2fb09216f8e2cc269bbfb8bb49b5|134213",
     ]
     default_url_template = "https://foobar.org/bin/v{version}/foobar-{version}-{platform}.tgz"
     default_url_platform_mapping = {
@@ -138,36 +140,86 @@ def no_exception():
 
 
 @pytest.mark.parametrize(
-    "version, assert_expectation",
+    "version, action, assert_expectation, expect_logged",
     [
         (
-            "1.2.3",
+            "v1.2.3",
+            ExternalToolVersionConstraintsViolationAction.RaiseError,
             pytest.raises(
                 UnsupportedVersion,
-                match="Version 1.2.3 does not satisfy the version constraint foobar<3.8,>3.2.1",
+                match="Version v1.2.3 does not satisfy the version constraints foobar<3.8,>3.2.1",
             ),
+            None,
         ),
         (
-            "3.2.2",
+            "v3.2.2",
+            ExternalToolVersionConstraintsViolationAction.RaiseError,
             pytest.raises(
-                UnknownVersion, match="No known version of foobar 3.2.2 for darwin found in"
+                UnknownVersion, match="No known version of foobar v3.2.2 for darwin found in"
             ),
+            None,
         ),
-        ("3.4.7", no_exception()),
         (
-            "3.8.0",
+            "v3.4.7",
+            ExternalToolVersionConstraintsViolationAction.RaiseError,
+            no_exception(),
+            None,
+        ),
+        (
+            "v3.8.0",
+            ExternalToolVersionConstraintsViolationAction.RaiseError,
             pytest.raises(
-                UnsupportedVersion, match="Version 3.8.0 does not satisfy the version constraint"
+                UnsupportedVersion, match="Version v3.8.0 does not satisfy the version constraints"
             ),
+            None,
+        ),
+        (
+            "v3.8.0",
+            ExternalToolVersionConstraintsViolationAction.LogWarning,
+            pytest.raises(
+                UnknownVersion, match="No known version of foobar v3.8.0 for darwin found in"
+            ),
+            [
+                (
+                    logging.WARNING,
+                    "foobar v3.8.0 may not be compatible with this version of pants, as it does not satisfy the constraints foobar<3.8,>3.2.1",
+                )
+            ],
+        ),
+        (
+            "v3.8.0",
+            ExternalToolVersionConstraintsViolationAction.Ignore,
+            pytest.raises(
+                UnknownVersion, match="No known version of foobar v3.8.0 for darwin found in"
+            ),
+            [
+                (
+                    logging.DEBUG,
+                    "Ignoring constraints for version v3.8.0, does not satisfy version constraints foobar<3.8,>3.2.1",
+                )
+            ],
         ),
     ],
 )
-def test_version_constraints(version, assert_expectation) -> None:
+def test_version_constraints(caplog, version, action, assert_expectation, expect_logged) -> None:
+    caplog.set_level(logging.DEBUG)
+    caplog.clear()
+
     with assert_expectation:
         create_subsystem(
             ConstrainedTool,
             version=version,
+            on_version_constraints_violation=action,
             known_versions=ConstrainedTool.default_known_versions,
             url_template=ConstrainedTool.default_url_template,
             url_platform_mapping=ConstrainedTool.default_url_platform_mapping,
         ).get_request(Platform.darwin)
+
+    if expect_logged:
+        assert len(caplog.records) == len(expect_logged)
+        for idx, (lvl, msg) in enumerate(expect_logged):
+            log_record = caplog.records[idx]
+            assert msg in log_record.message
+            assert lvl == log_record.levelno
+    else:
+        assert not caplog.records

--- a/src/python/pants/core/util_rules/external_tool_test.py
+++ b/src/python/pants/core/util_rules/external_tool_test.py
@@ -1,6 +1,8 @@
 # Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+from contextlib import contextmanager
+
 import pytest
 
 from pants.core.util_rules.external_tool import (
@@ -9,6 +11,7 @@ from pants.core.util_rules.external_tool import (
     ExternalToolRequest,
     TemplatedExternalTool,
     UnknownVersion,
+    UnsupportedVersion,
 )
 from pants.engine.fs import DownloadFile, FileDigest
 from pants.engine.platform import Platform
@@ -104,4 +107,67 @@ def test_generate_request() -> None:
     with pytest.raises(UnknownVersion):
         create_subsystem(
             FooBar, version="9.9.9", known_versions=FooBar.default_known_versions
+        ).get_request(Platform.darwin)
+
+
+class ConstrainedTool(TemplatedExternalTool):
+    name = "foobar"
+    options_scope = "foobar"
+    version_constraints = ">3.2.1, <3.8"
+    default_version = "3.4.7"
+    default_known_versions = [
+        "3.2.0|darwin   |1102324cdaacd589e50b8b7770595f220f54e18a1d76ee3c445198f80ab865b8|123346",
+        "3.2.0|linux_ppc|39e5d64b0f31117c94651c880d0a776159e49eab42b2066219569934b936a5e7|124443",
+        "3.2.0|linux    |c0c667fb679a8221bed01bffeed1f80727c6c7827d0cbd8f162195efb12df9e4|121212",
+        "3.4.7|darwin   |9d0e18cd74b918c7b3edd0203e75569e0c8caecb1367b3be409b45e28514f5be|123321",
+        "3.4.7|linux    |a019dfc4b32d63c1392aa264aed2253c1e0c2fb09216f8e2cc269bbfb8bb49b5|134213",
+    ]
+    default_url_template = "https://foobar.org/bin/v{version}/foobar-{version}-{platform}.tgz"
+    default_url_platform_mapping = {
+        "darwin": "osx-x86_64",
+        "linux": "linux-x86_64",
+    }
+
+    def generate_exe(self, plat: Platform) -> str:
+        return f"foobar-{self.version}/bin/foobar"
+
+
+@contextmanager
+def no_exception():
+    yield None
+
+
+@pytest.mark.parametrize(
+    "version, assert_expectation",
+    [
+        (
+            "1.2.3",
+            pytest.raises(
+                UnsupportedVersion,
+                match="Version 1.2.3 does not satisfy the version constraint foobar<3.8,>3.2.1",
+            ),
+        ),
+        (
+            "3.2.2",
+            pytest.raises(
+                UnknownVersion, match="No known version of foobar 3.2.2 for darwin found in"
+            ),
+        ),
+        ("3.4.7", no_exception()),
+        (
+            "3.8.0",
+            pytest.raises(
+                UnsupportedVersion, match="Version 3.8.0 does not satisfy the version constraint"
+            ),
+        ),
+    ],
+)
+def test_version_constraints(version, assert_expectation) -> None:
+    with assert_expectation:
+        create_subsystem(
+            ConstrainedTool,
+            version=version,
+            known_versions=ConstrainedTool.default_known_versions,
+            url_template=ConstrainedTool.default_url_template,
+            url_platform_mapping=ConstrainedTool.default_url_platform_mapping,
         ).get_request(Platform.darwin)

--- a/src/python/pants/core/util_rules/external_tool_test.py
+++ b/src/python/pants/core/util_rules/external_tool_test.py
@@ -2,7 +2,7 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
 import logging
-from contextlib import contextmanager
+import re
 
 import pytest
 
@@ -18,6 +18,7 @@ from pants.core.util_rules.external_tool import (
 from pants.engine.fs import DownloadFile, FileDigest
 from pants.engine.platform import Platform
 from pants.testutil.option_util import create_subsystem
+from pants.testutil.pytest_util import no_exception
 
 
 class FooBar(ExternalTool):
@@ -134,11 +135,6 @@ class ConstrainedTool(TemplatedExternalTool):
         return f"foobar-{self.version}/bin/foobar"
 
 
-@contextmanager
-def no_exception():
-    yield None
-
-
 @pytest.mark.parametrize(
     "version, action, assert_expectation, expect_logged",
     [
@@ -147,7 +143,11 @@ def no_exception():
             UnsupportedVersionUsage.RaiseError,
             pytest.raises(
                 UnsupportedVersion,
-                match="foobar v1.2.3 does not satisfy the version constraints foobar<3.8,>3.2.1",
+                match=re.escape(
+                    "The option [foobar].version is set to v1.2.3, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1\n"
+                    "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version.\n"
+                    "Alternatively, update [foobar].use_unsupported_version to be 'allow' or 'warning'."
+                ),
             ),
             None,
         ),
@@ -169,7 +169,12 @@ def no_exception():
             "v3.8.0",
             UnsupportedVersionUsage.RaiseError,
             pytest.raises(
-                UnsupportedVersion, match="foobar v3.8.0 does not satisfy the version constraints"
+                UnsupportedVersion,
+                match=re.escape(
+                    "The option [foobar].version is set to v3.8.0, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1\n"
+                    "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version.\n"
+                    "Alternatively, update [foobar].use_unsupported_version to be 'allow' or 'warning'."
+                ),
             ),
             None,
         ),
@@ -182,7 +187,10 @@ def no_exception():
             [
                 (
                     logging.WARNING,
-                    "foobar v3.8.0 may not be compatible with this version of pants, as it does not satisfy the constraints foobar<3.8,>3.2.1",
+                    (
+                        "The option [foobar].version is set to v3.8.0, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1\n"
+                        "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version."
+                    ),
                 )
             ],
         ),
@@ -194,8 +202,11 @@ def no_exception():
             ),
             [
                 (
-                    logging.DEBUG,
-                    "Ignoring constraints for version v3.8.0, does not satisfy version constraints foobar<3.8,>3.2.1",
+                    logging.INFO,
+                    (
+                        "The option [foobar].version is set to v3.8.0, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1\n"
+                        "The option [foobar].use_unsupported_version is set to 'allow'."
+                    ),
                 )
             ],
         ),

--- a/src/python/pants/core/util_rules/external_tool_test.py
+++ b/src/python/pants/core/util_rules/external_tool_test.py
@@ -144,9 +144,9 @@ class ConstrainedTool(TemplatedExternalTool):
             pytest.raises(
                 UnsupportedVersion,
                 match=re.escape(
-                    "The option [foobar].version is set to v1.2.3, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1\n"
-                    "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version.\n"
-                    "Alternatively, update [foobar].use_unsupported_version to be 'allow' or 'warning'."
+                    "The option [foobar].version is set to v1.2.3, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1. "
+                    "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version. "
+                    "Alternatively, update [foobar].use_unsupported_version to be 'warning'."
                 ),
             ),
             None,
@@ -171,9 +171,9 @@ class ConstrainedTool(TemplatedExternalTool):
             pytest.raises(
                 UnsupportedVersion,
                 match=re.escape(
-                    "The option [foobar].version is set to v3.8.0, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1\n"
-                    "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version.\n"
-                    "Alternatively, update [foobar].use_unsupported_version to be 'allow' or 'warning'."
+                    "The option [foobar].version is set to v3.8.0, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1. "
+                    "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version. "
+                    "Alternatively, update [foobar].use_unsupported_version to be 'warning'."
                 ),
             ),
             None,
@@ -188,24 +188,10 @@ class ConstrainedTool(TemplatedExternalTool):
                 (
                     logging.WARNING,
                     (
-                        "The option [foobar].version is set to v3.8.0, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1\n"
-                        "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version."
-                    ),
-                )
-            ],
-        ),
-        (
-            "v3.8.0",
-            UnsupportedVersionUsage.Allow,
-            pytest.raises(
-                UnknownVersion, match="No known version of foobar v3.8.0 for darwin found in"
-            ),
-            [
-                (
-                    logging.INFO,
-                    (
-                        "The option [foobar].version is set to v3.8.0, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1\n"
-                        "The option [foobar].use_unsupported_version is set to 'allow'."
+                        "The option [foobar].version is set to v3.8.0, which is not compatible with what this release of Pants expects: foobar<3.8,>3.2.1. "
+                        "Please update the version to a supported value, or consider using a different Pants release if you cannot change the version. "
+                        "Alternatively, you can ignore this warning (at your own peril) by adding this to the GLOBAL section of pants.toml: "
+                        'ignore_warnings = ["The option [foobar].version is set to"].'
                     ),
                 )
             ],

--- a/src/python/pants/testutil/pytest_util.py
+++ b/src/python/pants/testutil/pytest_util.py
@@ -1,0 +1,22 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from contextlib import contextmanager
+
+
+@contextmanager
+def no_exception():
+    """Useful replacement for `pytest.raises()`, when there is no exception to be expected.
+
+    When declaring parametrized tests, the test function can take a exceptions
+    expectation as input, and always use a with-block for the code under test.
+
+        @pytest.mark.parametrize('answer, expect_raises', [
+            (42, no_exception()),
+            (12, pytest.raises(WrongAnswer)),
+        ])
+        def test_search_for_the_meaning_of_life_universe_and_everything(answer, expect_raises):
+            with expect_raises:
+                computer.validate_result(answer)
+    """
+    yield None

